src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -365,6 +367,9 @@ public Next receive(Object o) {
             for (BodyExecutionCallback c : callbacks) {
                 c.onSuccess(sc, o);
             }
+            synchronized (this) {
+                thread.popContextVariables();
```

Comment:
I think the call to `popContextVariables` should be fast enough that there is no issue, but we could change it to the following if anyone prefers:

    CpsThread t;
    synchronized (this) {
        t = thread;
    }
    t.popContextVariables();